pkg/snapshot/restorer/restorer.go

```diff
@@ -887,3 +944,86 @@ func ErrorArrayToError(errs []error) error {
 
 	return fmt.Errorf("%s", strings.TrimSpace(errString))
 }
+
+// HandleAlarm function handles alarm raised by backup-restore.
+func (r *Restorer) HandleAlarm(stopHandleAlarmCh chan bool, dbSizeAlarmCh <-chan string, dbSizeDisAlarmCh chan bool, clientMaintenance client.MaintenanceCloser) {
+	r.logger.Info("Starting to handle an alarm...")
+	for {
+		select {
+		case <-stopHandleAlarmCh:
+			r.logger.Info("Closing handleAlarm...")
+			return
+		case endPoint := <-dbSizeAlarmCh:
+			r.logger.Info("Received a dbsize alarm")
+			r.logger.Infof("Calling defrag on endpoint: [%v]", endPoint)
+			if err := func() error {
+				ctx, cancel := context.WithTimeout(context.Background(), etcdDefragTimeout)
+				defer cancel()
+				if _, err := clientMaintenance.Defragment(ctx, endPoint); err != nil {
+					return err
+				}
+				return nil
+			}(); err != nil {
+				r.logger.Errorf("unable to disalarm as defrag call failed: %v", err)
+				// failed to disalarm
+				dbSizeDisAlarmCh <- false
+			} else {
+				// successfully disalarm
+				dbSizeDisAlarmCh <- true
+			}
+		}
+	}
+}
+
+// MakeEtcdLeanAndCheckAlarm calls etcd compaction on given revision number and raise db size alarm if embedded etcd db size crosses threshold.
+func (r *Restorer) MakeEtcdLeanAndCheckAlarm(revision int64, endPoints []string, embeddedEtcdQuotaBytes float64, dbSizeAlarmCh chan string, dbSizeDisAlarmCh <-chan bool, clientKV client.KVCloser, clientMaintenance client.MaintenanceCloser) error {
+	if err := func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), etcdCompactTimeout)
+		defer cancel()
+		if _, err := clientKV.Compact(ctx, revision, clientv3.WithCompactPhysical()); err != nil {
+			return fmt.Errorf("Compact API call failed: %v", err)
+		}
+		return nil
+	}(); err != nil {
+		return err
+	}
+	r.logger.Infof("Successfully compacted embedded etcd till revision: %v", revision)
+
+	ctx, cancel := context.WithTimeout(context.Background(), etcdConnectionTimeout)
+	defer cancel()
+
+	// check database size of embedded etcdServer.
+	status, err := clientMaintenance.Status(ctx, endPoints[0])
+	if err != nil {
+		return fmt.Errorf("unable to check embedded etcd status: %v", err)
+	}
+
+	if float64(status.DbSizeInUse) > thresholdPercentageForDBSizeAlarm*embeddedEtcdQuotaBytes ||
+		float64(status.DbSize) > thresholdPercentageForDBSizeAlarm*embeddedEtcdQuotaBytes {
```

Comment:
After compaction, trigger defragmentation based on the difference between `DbSizeInUse` and `DbSize`, instead of using a threshold size for `DbSizeInUse`, `DbSize`
WDYT @ishan16696  ?

Reply:
> based on the difference between DbSizeInUse and DbSize

lets assume `DbSize=7Gi` and `DbSizeInUse=6Gi`
then difference b/w `DbSize-DbSizeInUse=1Gi` then you still need some kind of benchmark(or threshold) to decide which way to go , isn't ?
And IMO difference b/w `DbSize` and `DbSizeInUse` can leads to some edge cases like when difference b/w 2 dbsize turned out to be `0` and so on, so why not just keep everything simple and use just a threshold with dbsize rather then some kind of difference which may leads to edge cases. WDYT ?